pkg/parcacol/ingest.go

```diff
@@ -18,211 +18,277 @@ import (
+	pprofpb "github.com/parca-dev/parca/gen/proto/go/google/pprof"
+	metastorepb "github.com/parca-dev/parca/gen/proto/go/parca/metastore/v1alpha1"
+	profilestorepb "github.com/parca-dev/parca/gen/proto/go/parca/profilestore/v1alpha1"
 	"github.com/parca-dev/parca/pkg/profile"
 )
 
+var ErrMissingNameLabel = errors.New("missing __name__ label")
+
 type Table interface {
 	Schema() *dynparquet.Schema
 	Insert(context.Context, []byte) (tx uint64, err error)
 }
 
-type Ingester struct {
+type NormalizedIngester struct {
 	logger     log.Logger
 	table      Table
-	normalizer *Normalizer
 	schema     *dynparquet.Schema
 	bufferPool *sync.Pool
+
+	allLabelNames         []string
+	allPprofLabelNames    []string
+	allPprofNumLabelNames []string
 }
 
-func NewIngester(
+func NewNormalizedIngester(
 	logger log.Logger,
-	normalizer *Normalizer,
 	table Table,
 	schema *dynparquet.Schema,
 	bufferPool *sync.Pool,
-) *Ingester {
-	return &Ingester{
+	allLabelNames []string,
+	allPprofLabelNames []string,
+	allPprofNumLabelNames []string,
+) NormalizedIngester {
+	return NormalizedIngester{
 		logger:     logger,
-		normalizer: normalizer,
 		table:      table,
 		schema:     schema,
 		bufferPool: bufferPool,
-	}
-}
-
-var ErrMissingNameLabel = errors.New("missing __name__ label")
-
-func separateNameFromLabels(ls labels.Labels) (string, map[string]struct{}, labels.Labels, error) {
-	names := make(map[string]struct{}, len(ls))
-	out := make(labels.Labels, 0, len(ls))
-	name := ""
-	for _, l := range ls {
-		if l.Name == "__name__" {
-			name = l.Value
-		} else {
-			names[l.Name] = struct{}{}
-			out = append(out, l)
-		}
-	}
 
-	if name == "" {
-		return "", nil, nil, ErrMissingNameLabel
+		allLabelNames:         allLabelNames,
+		allPprofLabelNames:    allPprofLabelNames,
+		allPprofNumLabelNames: allPprofNumLabelNames,
 	}
+}
 
-	return name, names, out, nil
+type Series struct {
+	Labels  map[string]string
+	Samples [][]*profile.NormalizedProfile
 }
 
-func (ing Ingester) Ingest(ctx context.Context, ls labels.Labels, p *pprofproto.Profile, normalized bool) error {
-	name, names, ls, err := separateNameFromLabels(ls)
+func (ing NormalizedIngester) Ingest(ctx context.Context, series []Series) error {
+	pBuf, err := ing.schema.GetBuffer(map[string][]string{
+		ColumnLabels:         ing.allLabelNames,
+		ColumnPprofLabels:    ing.allPprofLabelNames,
+		ColumnPprofNumLabels: ing.allPprofNumLabelNames,
+	})
 	if err != nil {
-		return fmt.Errorf("prepare labels: %w", err)
-	}
-	lset := ls.Map()
-
-	if err := validatePprofProfile(p); err != nil {
 		return err
 	}
+	defer ing.schema.PutBuffer(pBuf)
 
-	normalizedProfiles, err := ing.normalizer.NormalizePprof(ctx, name, names, p, normalized)
-	if err != nil {
-		return fmt.Errorf("normalize profile: %w", err)
-	}
+	var r parquet.Row
+	for _, s := range series {
+		for _, normalizedProfiles := range s.Samples {
+			for _, p := range normalizedProfiles {
```

Comment:
So this is the biggest change I would still propose to this.
Given that we built this nested `series>profiles>samples` construct earlier on, but then use it here for just a bit of reusing labels, I'd propose to flatten this earlier already.
Basically, the `Ingest` takes in `[]normalizedSamples` and we only have one loop to iterate over.